File: Univr Calendar/Core/ViewModels/CalendarViewModel.swift
//
//  CalendarViewModel.swift
//  Univr Calendar
//
//  Created by Leonardo Rossi on 22/10/25.
//

import Foundation

@Observable
class CalendarViewModel {
    var lessons: [Lesson] = []
    var days: [[Lesson]] = []
    var daysString: [String] = []
    
    var loading: Bool = true
    var checkingUpdates: Bool = false
    var showUpdateAlert: Bool = false
    var errorMessage: String? = nil
    var noLessonsFound: Bool = false
    
    private var pendingNewLessons: [Lesson]? = nil
    private var currentPalette: [String] = []
    private let cacheKey = "calendar_cache.json"
    
    private let networkService: NetworkServiceProtocol
    
    init(service: NetworkServiceProtocol = NetworkService()) {
        self.networkService = service
    }
    
    func loadFromCache(selYear: String, matricola: String) {
        if let cacheResponse = CacheManager.shared.load(fileName: cacheKey, type: ResponseAPI.self) {
            self.lessons = cacheResponse.celle
            self.organizeData(selectedYear: selYear, matricola: matricola)
            
            DispatchQueue.main.async {
                self.loading = false
                if self.lessons.isEmpty {
                    self.noLessonsFound = true
                }
            }
        }
    }
    
    @MainActor
    func loadLessons(corso: String, anno: String, selYear: String, matricola: String) async {
        guard corso != "0" else { return }
        
        if !lessons.isEmpty {
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.01) {
                self.checkingUpdates = true
            }
        }
        self.errorMessage = nil
        
        do {
            let response = try await networkService.fetchOrario(corso: corso, anno: anno, selyear: selYear)
            
            var fetchedLessons = response.celle
            self.currentPalette = response.colori
            
            if fetchedLessons.isEmpty {
                noLessonsFound = true
                self.applyNewData(fetchedLessons, palette: self.currentPalette, selectedYear: selYear, matricola: matricola)
            } else {
                assignColors(to: &fetchedLessons, palette: self.currentPalette)
                
                noLessonsFound = false
                if self.lessons.isEmpty {
                    self.applyNewData(fetchedLessons, palette: self.currentPalette, selectedYear: selYear, matricola: matricola)
                } else {
                    if !self.lessons.elementsEqual(fetchedLessons, by: {
                        $0.orario == $1.orario &&
                        $0.tipo == $1.tipo &&
                        $0.color == $1.color &&
                        $0.aula == $1.aula &&
                        $0.nomeInsegnamento == $1.nomeInsegnamento &&
                        $0.data == $1.data &&
                        $0.docente == $1.docente &&
                        $0.codiceInsegnamento == $1.codiceInsegnamento &&
                        $0.annullato == $1.annullato &&
                        $0.colorIndex == $1.colorIndex &&
                        $0.nameOriginal == $1.nameOriginal
                    }) {
                        self.pendingNewLessons = fetchedLessons
                        self.showUpdateAlert = true
                    }
                }
            }
        } catch {
            if let netError = error as? NetworkError {
                self.errorMessage = netError.localizedDescription
            } else {
                self.errorMessage = "Errore Generico: \(error.localizedDescription)"
            }
            print("Debug Error: \(error)")
        }
        
        if self.loading {
            self.loading = false
        }
        self.checkingUpdates = false
    }
    
    func clearPendingUpdate() {
        pendingNewLessons = nil
        checkingUpdates = false
        showUpdateAlert = false
    }
    
    func confirmUpdate(selectedYear: String, matricola: String) {
        guard let newLessons = pendingNewLessons else { return }
        
        applyNewData(newLessons, palette: self.currentPalette, selectedYear: selectedYear, matricola: matricola)
        
        noLessonsFound = false
        pendingNewLessons = nil
        showUpdateAlert = false
    }
    
    private func applyNewData(_ newLessons: [Lesson], palette: [String], selectedYear: String, matricola: String) {
        self.lessons = newLessons
        self.organizeData(selectedYear: selectedYear, matricola: matricola)
        
        let cacheObject = ResponseAPI(celle: newLessons, colori: palette)
        CacheManager.shared.save(cacheObject, fileName: cacheKey)
    }
    
    private func assignColors(to lessons: inout [Lesson], palette: [String]) {
        var colorMap: [String: String] = [:]
        var paletteIndex = 0
        
        for i in lessons.indices {
            if lessons[i].annullato == "1" {
                lessons[i].color = "#FFFFFF"
            } else if lessons[i].tipo == "chiusura_type" {
                lessons[i].color = "#BDF2F2"
            } else if !lessons[i].colorIndex.isEmpty {
               print("Trovato uno")
            } else if let existingColor = colorMap[lessons[i].codiceInsegnamento] {
                lessons[i].color = existingColor
            } else {
                let code = lessons[i].codiceInsegnamento
                
                if let existingColor = colorMap[code] {
                    lessons[i].color = existingColor
                } else {
                    let newColor = (paletteIndex < palette.count) ? palette[paletteIndex] : "#CCCCCC"
                    colorMap[code] = newColor
                    lessons[i].color = newColor
                    
                    paletteIndex += 1
                }
            }
        }
    }
    
    func organizeData(selectedYear: String, matricola: String) {
        guard let annoInt = Int(selectedYear) else { return }
        
        self.daysString = generaDateAnnoAccademico(annoInizio: annoInt)
        var newDays: [[Lesson]] = []
        
        for day in daysString {
            var filtered = lessons.filter { lesson in
                let isRightDay = lesson.data == day
                let isNotChiusura = (lesson.tipo != "chiusura_type")
                
                let nameCondition = lesson.nomeInsegnamento.contains("Matricole \(matricola)") || (!lesson.nomeInsegnamento.contains("Matricole pari") && !lesson.nomeInsegnamento.contains("Matricole dispari"))
                
                return isRightDay && isNotChiusura && nameCondition
            }
            
            filtered = filtered.sorted(by: { $0.orario < $1.orario })
            
            if !filtered.isEmpty {
                var processedDay = filtered
                var added = 0
                
                for i in filtered.indices {
                    if i + 1 < filtered.count {
                        let endString = String(filtered[i].orario.split(separator: " - ").last ?? "")
                        let startString = String(filtered[i + 1].orario.split(separator: " - ").first ?? "")
                        
                        let endDate = endString.date(format: "HH:mm")
                        let startDate = startString.date(format: "HH:mm")
                        
                        if let end = endDate, let start = startDate, end.timeIntervalSince1970 < start.timeIntervalSince1970 {
                            
                            processedDay.insert(Lesson(
                                data: day,
                                orario: endString + " - " + startString,
                                tipo: "pause"
                            ), at: i + added + 1)
                            added += 1
                        }
                    }
                }
                newDays.append(processedDay)
            } else {
                newDays.append(filtered)
            }
        }
        
        self.days = newDays
    }
    
    func generaDateAnnoAccademico(annoInizio: Int) -> [String] {
        var dateStringhe: [String] = []
        let calendar = Calendar.current
        
        let formatter = Formatters.displayDate
        
        var startComponents = DateComponents()
        startComponents.year = annoInizio
        startComponents.month = 10
        startComponents.day = 1
        
        var endComponents = DateComponents()
        endComponents.year = annoInizio + 1
        endComponents.month = 9
        endComponents.day = 30
        
        guard let startDate = calendar.date(from: startComponents),
              let endDate = calendar.date(from: endComponents) else {
            return []
        }
        
        var currentDate = startDate
        
        while currentDate <= endDate {
            dateStringhe.append(formatter.string(from: currentDate))
            if let nextDate = calendar.date(byAdding: .day, value: 1, to: currentDate) {
                currentDate = nextDate
            } else {
                break
            }
        }
        
        return dateStringhe
    }
}

File: Android/core/src/main/java/dev/leonardorossi/univrcore/CalendarViewModel.kt
//
//  CalendarViewModel.kt
//  UnivrCore
//
//  Translated from CalendarViewModel.swift
//  Copyright (C) 2026 Leonardo Rossi
//  SPDX-License-Identifier: GPL-3.0-or-later
//

package dev.leonardorossi.univrcore

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.coroutines.yield
import java.time.LocalDate

internal data class YearStructure(
    val year: Int,
    val days: List<String>,
    val dates: List<LocalDate>
)

class CalendarViewModel(
    private val cacheManager: CacheManager
) : ViewModel() {

    private val _lessons = MutableStateFlow<List<Lesson>>(emptyList())
    val lessons: StateFlow<List<Lesson>> = _lessons.asStateFlow()

    private val _days = MutableStateFlow<List<List<Lesson>>>(emptyList())
    val days: StateFlow<List<List<Lesson>>> = _days.asStateFlow()

    private val _daysString = MutableStateFlow<List<String>>(emptyList())
    val daysString: StateFlow<List<String>> = _daysString.asStateFlow()

    private val _loading = MutableStateFlow(true)
    val loading: StateFlow<Boolean> = _loading.asStateFlow()

    private val _checkingUpdates = MutableStateFlow(false)
    val checkingUpdates: StateFlow<Boolean> = _checkingUpdates.asStateFlow()

    private val _showUpdateAlert = MutableStateFlow(false)
    val showUpdateAlert: StateFlow<Boolean> = _showUpdateAlert.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    private val _noLessonsFound = MutableStateFlow(false)
    val noLessonsFound: StateFlow<Boolean> = _noLessonsFound.asStateFlow()

    private val _isOffline = MutableStateFlow(false)
    val isOffline: StateFlow<Boolean> = _isOffline.asStateFlow()

    private var pendingNewLessons: List<Lesson>? = null
    private var currentPalette: List<String> = emptyList()
    private var cachedStructure: YearStructure? = null

    private val service = NetworkService()
    private val cacheKey = "calendar_cache.json"

    suspend fun loadFromCache(selYear: String, matricola: String) {
        val cacheResponse = cacheManager.load(cacheKey, ResponseAPI.serializer())
        if (cacheResponse != null) {
            _lessons.value = cacheResponse.celle
            currentPalette = cacheResponse.colori

            organizeData(selectedYear = selYear, matricola = matricola)

            _loading.value = false
            _noLessonsFound.value = _lessons.value.isEmpty()
        }
    }

    suspend fun loadNetworkFromCache() {
        val cacheResponse = cacheManager.load("network_cache.json", NetworkCacheData.serializer())
        if (cacheResponse != null) {
            NetworkCache.update(cacheResponse)
        }
    }

    fun loadLessons(corso: String, anno: String, selYear: String, matricola: String, updating: Boolean) {
        if (corso == "0") return
        _isOffline.value = false

        viewModelScope.launch {
            if (!updating && _lessons.value.isEmpty()) {
                loadFromCache(selYear, matricola)
            }

            if (!updating && _lessons.value.isNotEmpty()) {
                _checkingUpdates.value = true
            } else {
                clearAll()
                _loading.value = true
            }
            _errorMessage.value = null

            try {
                val response = service.fetchOrario(corso = corso, anno = anno, selyear = selYear)

                currentPalette = response.colori

                var fetchedLessons = response.celle
                if (fetchedLessons.isNotEmpty()) {
                    fetchedLessons = CalendarLogic.applyColors(fetchedLessons, currentPalette)
                }

                handleNewData(fetchedLessons, selYear, matricola, updating)

                _loading.value = false
            } catch (e: Exception) {
                handleError(e)
            }

            _checkingUpdates.value = false
        }
    }

    fun confirmUpdate(selectedYear: String, matricola: String) {
        viewModelScope.launch {
            val newLessons = pendingNewLessons ?: return@launch

            try {
                handleNewData(newLessons, selectedYear, matricola, update = true)
            } catch (_: Exception) { }

            _noLessonsFound.value = false
            pendingNewLessons = null
            _showUpdateAlert.value = false
        }
    }

    fun clearPendingUpdate() {
        pendingNewLessons = null
        _checkingUpdates.value = false
        _showUpdateAlert.value = false
    }

    suspend fun organizeData(selectedYear: String, matricola: String) {
        val annoInt = selectedYear.toIntOrNull() ?: return

        val lessonsSnapshot = _lessons.value
        val currentCache = cachedStructure

        val (newStructure, organizedDays) = CalendarLogic.processCalendarData(
            year = annoInt,
            matricola = matricola,
            lessons = lessonsSnapshot,
            cachedStructure = currentCache
        )

        if (cachedStructure?.year != annoInt) {
            cachedStructure = newStructure
        }
        _daysString.value = newStructure.days
        _days.value = organizedDays

        val activeActivities = mutableMapOf<String, Double>()
        for (index in organizedDays.indices) {
            val date = newStructure.days[index]
            val dayLessons = organizedDays[index]

            val validLessons = dayLessons.filter { !it.annullato && it.tipo != "pause" }

            if (validLessons.isNotEmpty()) {
                val totalMinutes = validLessons.sumOf { lesson ->
                    val times = lesson.orario.split("-").map { it.trim() }
                    if (times.size != 2) return@sumOf 0

                    fun toMinutes(time: String): Int {
                        val parts = time.split(":")
                        if (parts.size != 2) return 0
                        val h = parts[0].toIntOrNull() ?: return 0
                        val m = parts[1].toIntOrNull() ?: return 0
                        return h * 60 + m
                    }

                    val start = toMinutes(times[0])
                    val end = toMinutes(times[1])
                    end - start
                }

                val hours = totalMinutes / 60.0
                if (hours > 0) {
                    activeActivities[date] = hours
                }
            }
        }

        DatePickerCache.getInstance().updateActivities(activeActivities)
    }

    private suspend fun handleNewData(
        fetchedLessons: List<Lesson>,
        selectedYear: String,
        matricola: String,
        update: Boolean
    ) {
        if (fetchedLessons.isEmpty()) {
            _noLessonsFound.value = true
            updateStateAndCache(emptyList(), selectedYear, matricola)
            return
        }

        _noLessonsFound.value = false

        if (_lessons.value.isEmpty() || update) {
            updateStateAndCache(fetchedLessons, selectedYear, matricola)
            return
        }

        if (_lessons.value != fetchedLessons) {
            pendingNewLessons = fetchedLessons
            _showUpdateAlert.value = true
        }
    }

    private suspend fun updateStateAndCache(
        newLessons: List<Lesson>,
        selectedYear: String,
        matricola: String
    ) {
        _lessons.value = newLessons

        val cacheObject = ResponseAPI(celle = newLessons, colori = currentPalette)
        cacheManager.save(cacheObject, cacheKey, ResponseAPI.serializer())

        organizeData(selectedYear, matricola)
    }

    private suspend fun clearCache() {
        cacheManager.clear(cacheKey)
    }

    suspend fun clearAll() {
        _loading.value = true
        yield()
        clearCache()
        clearPendingUpdate()
        _lessons.value = emptyList()
        _days.value = emptyList()
    }

    private fun handleError(error: Exception) {
        if (error is NetworkError) {
            if (error is NetworkError.Offline) {
                _isOffline.value = _lessons.value.isEmpty()
            }
            _errorMessage.value = error.message
        } else {
            _errorMessage.value = "Errore generico: ${error.localizedMessage}"
        }
        println("Debug Error: $error")
    }
}

// ── CalendarLogic (pure functions) ──

internal object CalendarLogic {

    suspend fun processCalendarData(
        year: Int,
        matricola: String,
        lessons: List<Lesson>,
        cachedStructure: YearStructure?
    ): Pair<YearStructure, List<List<Lesson>>> = withContext(Dispatchers.Default) {
        val structure: YearStructure = if (cachedStructure != null && cachedStructure.year == year) {
            cachedStructure
        } else {
            generateYearStructure(year)
        }

        val lessonsByDate = lessons.groupBy { it.data }
        val userFilter = if (matricola == "pari") Lesson.GruppoMatricola.PARI else Lesson.GruppoMatricola.DISPARI

        val organized = ArrayList<List<Lesson>>(structure.days.size)

        for (dayString in structure.days) {
            val dailyLessons = lessonsByDate[dayString]
            if (dailyLessons == null) {
                organized.add(emptyList())
                continue
            }

            val filtered = dailyLessons.filter { lesson ->
                lesson.tipo != "chiusura_type" &&
                        (lesson.gruppo == Lesson.GruppoMatricola.TUTTI || lesson.gruppo == userFilter)
            }.sortedBy { it.orario }

            if (filtered.isEmpty()) {
                organized.add(emptyList())
            } else {
                organized.add(insertPauses(filtered, dayString))
            }
        }

        Pair(structure, organized)
    }

    fun generateYearStructure(year: Int): YearStructure {
        val startDate = LocalDate.of(year, 10, 1)
        val endDate = LocalDate.of(year + 1, 9, 30)

        val dateStrings = mutableListOf<String>()
        val dateObj = mutableListOf<LocalDate>()

        var currentDate = startDate
        while (!currentDate.isAfter(endDate)) {
            dateStrings.add(currentDate.formatUnivrStyle())
            dateObj.add(currentDate)
            currentDate = currentDate.plusDays(1)
        }

        return YearStructure(year = year, days = dateStrings, dates = dateObj)
    }

    private fun insertPauses(lessons: List<Lesson>, date: String): List<Lesson> {
        val processedDay = lessons.toMutableList()
        var offset = 0

        for (i in 0 until lessons.size - 1) {
            val currentEnd = lessons[i].orario.takeLast(5)
            val nextStart = lessons[i + 1].orario.take(5)

            if (currentEnd < nextStart) {
                val pauseLesson = Lesson(
                    data = date,
                    orario = "$currentEnd-$nextStart",
                    tipo = "pause"
                )
                processedDay.add(i + 1 + offset, pauseLesson)
                offset++
            }
        }
        return processedDay
    }

    fun applyColors(lessons: List<Lesson>, palette: List<String>): List<Lesson> {
        val processedLessons = lessons.map { it.copy() }.toMutableList()
        val colorMap = mutableMapOf<String, String>()
        var paletteIndex = 0

        for (lesson in processedLessons) {
            if (hasCustomColor(lesson)) {
                colorMap[lesson.codiceInsegnamento] = lesson.color
            }
        }

        for (i in processedLessons.indices) {
            if (processedLessons[i].annullato) {
                processedLessons[i] = processedLessons[i].copy(color = "#FFFFFF")
                continue
            }

            if (processedLessons[i].tipo == "chiusura_type") {
                processedLessons[i] = processedLessons[i].copy(color = "#BDF2F2")
                continue
            }

            if (processedLessons[i].colorIndex.isNotEmpty()) {
                println("Trovato uno")
                continue
            }

            val code = lessons[i].codiceInsegnamento

            val existingColor = colorMap[code]
            if (existingColor != null) {
                processedLessons[i] = processedLessons[i].copy(color = existingColor)
            } else {
                val newColor = if (paletteIndex < palette.size) palette[paletteIndex] else "#CCCCCC"
                colorMap[code] = newColor
                processedLessons[i] = processedLessons[i].copy(color = newColor)
                paletteIndex++
            }
        }
        return processedLessons
    }

    private fun hasCustomColor(lesson: Lesson): Boolean {
        return lesson.color.isNotEmpty() && lesson.color != "CCCCCC" && lesson.color != "A0A0A0"
    }
}
